### PR TITLE
Consumer diagnostics

### PR DESCRIPTION
**Pull request by andreamarcolin**
Implements #21 

- Added DiagnosticEvent ADT
- Added Diagnostics interface
- Added no-op implementation of Diagnostics service
- Added sliding queue implementation of Diagnostics service
- Implemented diagnostics emission via a Diagnostics service (part of Runloop Deps, defaults to no-op implementation on consumer constructor) throughout the Runloop

### CODE REVIEW

**Comment by andreamarcolin** [comment]
Rebased on top of master after 0.2.0 and resolved conflicts

File: src/main/scala/zio/kafka/client/Runloop.scala [hunk 1]
```diff
@@ -45,7 +46,8 @@ object Runloop {
     partitions: Queue[Take[Throwable, (TopicPartition, ZStreamChunk[Any, Throwable, ByteArrayCommittableRecord])]],
     rebalances: Queue[Rebalance],
```
Comment (iravid) [inline src/main/scala/zio/kafka/client/Runloop.scala]:
```suggestion

```

File: src/main/scala/zio/kafka/client/Runloop.scala [hunk 2]
```diff
@@ -76,16 +85,24 @@ object Runloop {
                        new ConsumerRebalanceListener {
                          override def onPartitionsRevoked(partitions: java.util.Collection[TopicPartition]): Unit = {
                            runtime.unsafeRun(
-                             rebalancingRef
-                               .set(true) *> rebalances.offer(Rebalance.Revoked(partitions.asScala.toSet)).fork
+                             for {
+                               _       <- rebalancingRef.set(true)
+                               revoked = partitions.asScala.toSet
+                               handle  <- rebalances.offer(Rebalance.Revoked(revoked)).fork
```
Comment (iravid) [inline src/main/scala/zio/kafka/client/Runloop.scala]:
```suggestion

```

File: src/main/scala/zio/kafka/client/Runloop.scala [hunk 3]
```diff
@@ -76,16 +85,24 @@ object Runloop {
...
                          }
 
                          override def onPartitionsAssigned(partitions: java.util.Collection[TopicPartition]): Unit = {
                            runtime.unsafeRun(
-                             rebalancingRef
-                               .set(false) *> rebalances.offer(Rebalance.Assigned(partitions.asScala.toSet)).fork
+                             for {
+                               _        <- rebalancingRef.set(false)
+                               assigned = partitions.asScala.toSet
+                               handle   <- rebalances.offer(Rebalance.Assigned(partitions.asScala.toSet)).fork
```
Comment (iravid) [inline src/main/scala/zio/kafka/client/Runloop.scala]:
```suggestion

```

File: src/main/scala/zio/kafka/client/diagnostics/DiagnosticEvent.scala [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+object DiagnosticEvent {
+  case class Poll(tpRequested: Set[TopicPartition], tpWithData: Set[TopicPartition], tpWithoutData: Set[TopicPartition])
+      extends DiagnosticEvent
+  case class Request(partition: TopicPartition) extends DiagnosticEvent
+
+  sealed trait Commit extends DiagnosticEvent
+  object Commit {
+    case class Started(offsets: Map[TopicPartition, Long])              extends Commit
+    case class Success(offsets: Map[TopicPartition, OffsetAndMetadata]) extends Commit
+    case class Failure(offsets: Map[TopicPartition, OffsetAndMetadata]) extends Commit
```
Comment (iravid) [inline src/main/scala/zio/kafka/client/diagnostics/DiagnosticEvent.scala]: Let's add a `cause: Throwable`?

File: src/main/scala/zio/kafka/client/diagnostics/Diagnostics.scala [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package zio.kafka.client.diagnostics
+
+import zio.{ Managed, Queue, UIO }
+
+trait Diagnostics {
+  def emit(event: DiagnosticEvent): UIO[Unit]
```
Comment (iravid) [inline src/main/scala/zio/kafka/client/diagnostics/Diagnostics.scala]: Let's add a `val enabled: Boolean` here, and check that before we allocate events and emit them. The no-op implementation can then really be a no-op implementation.

Possibly a helper `def emitIfEnabled(event: => DiagnosticEvent)` could come in handy.

File: src/main/scala/zio/kafka/client/diagnostics/DiagnosticEvent.scala (hunk 4 above)
Comment (andreamarcolin) [reply src/main/scala/zio/kafka/client/diagnostics/DiagnosticEvent.scala]: done

File: src/main/scala/zio/kafka/client/Runloop.scala (hunk 2 above)
Comment (andreamarcolin) [reply src/main/scala/zio/kafka/client/Runloop.scala]: removed

File: src/main/scala/zio/kafka/client/Runloop.scala (hunk 3 above)
Comment (andreamarcolin) [reply src/main/scala/zio/kafka/client/Runloop.scala]: removed

File: src/main/scala/zio/kafka/client/Runloop.scala (hunk 1 above)
Comment (andreamarcolin) [reply src/main/scala/zio/kafka/client/Runloop.scala]: removed

File: src/main/scala/zio/kafka/client/diagnostics/Diagnostics.scala (hunk 5 above)
Comment (andreamarcolin) [reply src/main/scala/zio/kafka/client/diagnostics/Diagnostics.scala]: Implemented as suggested and ready for another review

**Pull request merged**